File: actin-common/src/test/java/com/hartwig/actin/clinical/feed/TestFeedFactory.java
package com.hartwig.actin.clinical.feed;

import java.time.LocalDate;
import java.util.List;

import com.google.common.collect.Lists;
import com.hartwig.actin.clinical.datamodel.Sex;
import com.hartwig.actin.clinical.feed.bloodpressure.BloodPressureEntry;
import com.hartwig.actin.clinical.feed.complication.ComplicationEntry;
import com.hartwig.actin.clinical.feed.encounter.EncounterEntry;
import com.hartwig.actin.clinical.feed.encounter.ImmutableEncounterEntry;
import com.hartwig.actin.clinical.feed.intolerance.ImmutableIntoleranceEntry;
import com.hartwig.actin.clinical.feed.intolerance.IntoleranceEntry;
import com.hartwig.actin.clinical.feed.lab.ImmutableLabEntry;
import com.hartwig.actin.clinical.feed.lab.LabEntry;
import com.hartwig.actin.clinical.feed.medication.MedicationEntry;
import com.hartwig.actin.clinical.feed.patient.ImmutablePatientEntry;
import com.hartwig.actin.clinical.feed.patient.PatientEntry;
import com.hartwig.actin.clinical.feed.questionnaire.ImmutableQuestionnaireEntry;
import com.hartwig.actin.clinical.feed.questionnaire.QuestionnaireEntry;
import com.hartwig.actin.clinical.feed.questionnaire.TestQuestionnaireFactory;

import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;

public final class TestFeedFactory {

    static final String TEST_SUBJECT = "ACTN-01-02-9999";

    private TestFeedFactory() {
    }

    @NotNull
    public static FeedModel createProperTestFeedModel() {
        return new FeedModel(createTestClinicalFeed());
    }

    @NotNull
    public static FeedModel createMinimalTestFeedModel() {
        return new FeedModel(ImmutableClinicalFeed.builder().patientEntries(createTestPatientEntries()).build());
    }

    @NotNull
    public static ClinicalFeed createTestClinicalFeed() {
        return ImmutableClinicalFeed.builder()
                .patientEntries(createTestPatientEntries())
                .questionnaireEntries(createTestQuestionnaireEntries())
                .encounterEntries(createTestEncounterEntries())
                .medicationEntries(createTestMedicationEntries())
                .labEntries(createTestLabEntries())
                .bloodPressureEntries(createTestBloodPressureEntries())
                .complicationEntries(createTestComplicationEntries())
                .intoleranceEntries(createTestIntoleranceEntries())
                .build();
    }

    @NotNull
    private static List<PatientEntry> createTestPatientEntries() {
        return Lists.newArrayList(ImmutablePatientEntry.builder()
                .id("ID")
                .subject(TEST_SUBJECT)
                .birthYear(1960)
                .sex(Sex.MALE)
                .periodStart(LocalDate.of(2021, 6, 1))
                .periodEnd(LocalDate.of(2021, 10, 1))
                .build());
    }

    @NotNull
    private static List<QuestionnaireEntry> createTestQuestionnaireEntries() {
        ImmutableQuestionnaireEntry.Builder questionnaireBuilder =
                ImmutableQuestionnaireEntry.builder().from(TestQuestionnaireFactory.createTestQuestionnaireEntry());

        List<QuestionnaireEntry> entries = Lists.newArrayList();
        entries.add(questionnaireBuilder.subject(TEST_SUBJECT).authoredDateTime(LocalDate.of(2021, 7, 1)).build());
        entries.add(questionnaireBuilder.subject(TEST_SUBJECT).authoredDateTime(LocalDate.of(2021, 8, 1)).build());

        ImmutableQuestionnaireEntry.Builder toxicityBuilder = ImmutableQuestionnaireEntry.builder()
                .subject(TEST_SUBJECT)
                .parentIdentifierValue(Strings.EMPTY)
                .authoredDateTime(LocalDate.of(2020, 6, 6))
                .questionnaireQuestionnaireValue(Strings.EMPTY)
                .description("ONC Kuuroverzicht");

        entries.add(toxicityBuilder.itemText("Nausea").itemAnswerValueValueString("2").build());
        entries.add(toxicityBuilder.itemText("Vomiting").itemAnswerValueValueString(Strings.EMPTY).build());

        return entries;
    }

    @NotNull
    private static List<EncounterEntry> createTestEncounterEntries() {
        List<EncounterEntry> entries = Lists.newArrayList();

        entries.add(ImmutableEncounterEntry.builder()
                .subject(TEST_SUBJECT)
                .type1Display(Strings.EMPTY)
                .classDisplay("surgery")
                .periodStart(LocalDate.of(2015, 10, 10))
                .periodEnd(LocalDate.of(2015, 10, 10))
                .identifierValue("ID")
                .identifierSystem("URL")
                .codeCodingCodeOriginal("code")
                .codeCodingDisplayOriginal("diagnostics")
                .presentedFormData(Strings.EMPTY)
                .reason(Strings.EMPTY)
                .accessionValue(Strings.EMPTY)
                .build());

        return entries;
    }

    @NotNull
    private static List<MedicationEntry> createTestMedicationEntries() {
        return Lists.newArrayList();
    }

    @NotNull
    public static List<LabEntry> createTestLabEntries() {
        List<LabEntry> entries = Lists.newArrayList();

        ImmutableLabEntry.Builder baseBuilder = ImmutableLabEntry.builder()
                .subject(TEST_SUBJECT)
                .valueQuantityComparator(Strings.EMPTY)
                .valueString(Strings.EMPTY)
                .codeCode(Strings.EMPTY);

        entries.add(baseBuilder.codeCodeOriginal("LAB1")
                .codeDisplayOriginal("Lab Value 1")
                .issued(LocalDate.of(2018, 5, 29))
                .valueQuantityValue(30D)
                .valueQuantityUnit("U/l")
                .interpretationDisplayOriginal("ok")
                .referenceRangeText("20 - 40")
                .build());

        entries.add(baseBuilder.codeCodeOriginal("LAB2")
                .codeDisplayOriginal("Lab Value 2")
                .issued(LocalDate.of(2018, 5, 29))
                .valueQuantityValue(22D)
                .valueQuantityUnit("mmol/l")
                .interpretationDisplayOriginal("too low")
                .referenceRangeText("> 30")
                .build());

        entries.add(baseBuilder.codeCodeOriginal("LAB3")
                .codeDisplayOriginal("Lab Value 3")
                .issued(LocalDate.of(2018, 5, 29))
                .valueQuantityComparator(">")
                .valueQuantityValue(50D)
                .valueQuantityUnit("mL/min")
                .interpretationDisplayOriginal("ok")
                .referenceRangeText("> 50")
                .build());

        entries.add(baseBuilder.codeCodeOriginal("LAB4")
                .codeDisplayOriginal("Lab Value 4")
                .issued(LocalDate.of(2018, 5, 29))
                .valueQuantityValue(20D)
                .valueQuantityUnit("mL/min")
                .interpretationDisplayOriginal("ok")
                .referenceRangeText(Strings.EMPTY)
                .build());

        return entries;
    }

    @NotNull
    private static List<BloodPressureEntry> createTestBloodPressureEntries() {
        return Lists.newArrayList();
    }

    @NotNull
    private static List<ComplicationEntry> createTestComplicationEntries() {
        return Lists.newArrayList();
    }

    @NotNull
    private static List<IntoleranceEntry> createTestIntoleranceEntries() {
        List<IntoleranceEntry> entries = Lists.newArrayList();

        entries.add(ImmutableIntoleranceEntry.builder()
                .subject(TEST_SUBJECT)
                .assertedDate(LocalDate.of(2018, 1, 1))
                .category("medication")
                .categoryAllergyCategoryCode(Strings.EMPTY)
                .categoryAllergyCategoryDisplay(Strings.EMPTY)
                .clinicalStatus(Strings.EMPTY)
                .codeText("pills")
                .criticality("unknown")
                .build());

        return entries;
    }
}

File: clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/emc/TestFeedFactory.kt
package com.hartwig.actin.clinical.feed.emc

import com.hartwig.actin.clinical.curation.FULL_ATC_CODE
import com.hartwig.actin.clinical.feed.emc.bodyweight.BodyWeightEntry
import com.hartwig.actin.clinical.feed.emc.digitalfile.DigitalFileEntry
import com.hartwig.actin.clinical.feed.emc.intolerance.IntoleranceEntry
import com.hartwig.actin.clinical.feed.emc.lab.LabEntry
import com.hartwig.actin.clinical.feed.emc.medication.MedicationEntry
import com.hartwig.actin.clinical.feed.emc.patient.PatientEntry
import com.hartwig.actin.clinical.feed.emc.questionnaire.QuestionnaireEntry
import com.hartwig.actin.clinical.feed.emc.questionnaire.TestQuestionnaireFactory
import com.hartwig.actin.clinical.feed.emc.surgery.SurgeryEntry
import com.hartwig.actin.clinical.feed.emc.vitalfunction.VitalFunctionEntry
import com.hartwig.actin.datamodel.clinical.Gender
import com.hartwig.actin.testutil.ResourceLocator.resourceOnClasspath
import java.time.LocalDate
import java.time.LocalDateTime

private const val TEST_SUBJECT = "ACTN-01-02-9999"
val FEED_DIRECTORY = resourceOnClasspath("feed/emc") + "/"

object TestFeedFactory {

    fun createProperTestFeedModel(): FeedModel {
        return FeedModel(createTestClinicalFeed())
    }

    fun createTestClinicalFeed(): EmcClinicalFeed {
        return EmcClinicalFeed(
            patientEntries = createTestPatientEntries(),
            questionnaireEntries = createTestQuestionnaireEntries(),
            digitalFileEntries = createTestDigitalFileEntries(),
            surgeryEntries = createTestSurgeryEntries(),
            medicationEntries = createTestMedicationEntries(),
            labEntries = createTestLabEntries(),
            vitalFunctionEntries = createTestVitalFunctionEntries(),
            intoleranceEntries = createTestIntoleranceEntries(),
            bodyWeightEntries = createTestBodyWeightEntries()
        )
    }

    private fun createTestPatientEntries(): List<PatientEntry> {
        return listOf(
            PatientEntry(
                subject = TEST_SUBJECT,
                birthYear = 1960,
                gender = Gender.MALE,
                periodStart = LocalDate.of(2021, 6, 1),
                periodEnd = LocalDate.of(2021, 10, 1)
            )
        )
    }

    private fun createTestQuestionnaireEntries(): List<QuestionnaireEntry> {
        val questionnaireBase = TestQuestionnaireFactory.createTestQuestionnaireEntry()
        return listOf(
            questionnaireBase.copy(subject = TEST_SUBJECT, authored = LocalDate.of(2021, 7, 1)),
            questionnaireBase.copy(subject = TEST_SUBJECT, authored = LocalDate.of(2021, 8, 1))
        )
    }

    private fun createTestDigitalFileEntries(): List<DigitalFileEntry> {
        val minimalToxicity = minimalDigitalFileEntry(LocalDate.of(2020, 6, 6), "ONC Kuuroverzicht")
        return listOf(
            minimalToxicity.copy(itemText = "Nausea", itemAnswerValueValueString = "2"),
            minimalToxicity.copy(itemText = "Vomiting", itemAnswerValueValueString = ""),
            minimalToxicity.copy(itemText = "Pain", itemAnswerValueValueString = "0. Not applicable"),
            minimalDigitalFileEntry(LocalDate.of(2020, 7, 7), "Aanvraag bloedproducten_test").copy(itemAnswerValueValueString = "Product")
        )
    }

    private fun minimalDigitalFileEntry(authored: LocalDate, description: String): DigitalFileEntry {
        return DigitalFileEntry(
            subject = TEST_SUBJECT,
            authored = authored,
            description = description,
            itemText = "",
            itemAnswerValueValueString = ""
        )
    }

    private fun createTestSurgeryEntries(): List<SurgeryEntry> {
        return listOf(
            surgeryEntry(LocalDate.of(2015, 10, 10), LocalDate.of(2015, 10, 10)),
            surgeryEntry(LocalDate.of(2015, 10, 10), LocalDate.of(2015, 10, 10))
        )
    }

    private fun surgeryEntry(periodStart: LocalDate, periodEnd: LocalDate): SurgeryEntry {
        return SurgeryEntry(
            subject = TEST_SUBJECT,
            classDisplay = "surgery",
            codeCodingDisplayOriginal = "diagnostics",
            encounterStatus = "planned",
            procedureStatus = "planned",
            periodStart = periodStart,
            periodEnd = periodEnd
        )
    }

    private fun createTestMedicationEntries(): List<MedicationEntry> {
        return listOf(
            medicationEntry(
                status = "active",
                dosageInstruction = "once per day 50-60 mg every month",
                start = LocalDate.of(2019, 2, 2),
                end = LocalDate.of(2019, 4, 4),
                active = true,
                code5ATCCode = FULL_ATC_CODE,
                code5ATCDisplay = "PARACETAMOL",
                administrationRoute = "oraal"
            ), medicationEntry(
                status = "",
                dosageInstruction = "Irrelevant",
                start = LocalDate.of(2019, 2, 2),
                end = LocalDate.of(2019, 4, 4),
                active = false,
                code5ATCDisplay = ""
            )
        )
    }

    fun medicationEntry(
        status: String, dosageInstruction: String, start: LocalDate, end: LocalDate, active: Boolean,
        code5ATCDisplay: String = "",
        administrationRoute: String = "",
        code5ATCCode: String = "",
        codeText: String = ""
    ): MedicationEntry {
        return MedicationEntry(
            status = status,
            dosageInstructionText = dosageInstruction,
            periodOfUseValuePeriodStart = start,
            periodOfUseValuePeriodEnd = end,
            active = active,
            code5ATCDisplay = code5ATCDisplay,
            subject = TEST_SUBJECT,
            codeText = codeText,
            code5ATCCode = code5ATCCode,
            chemicalSubgroupDisplay = "",
            pharmacologicalSubgroupDisplay = "",
            therapeuticSubgroupDisplay = "",
            anatomicalMainGroupDisplay = "",
            dosageInstructionRouteDisplay = administrationRoute,
            dosageInstructionDoseQuantityUnit = "",
            dosageInstructionDoseQuantityValue = 0.0,
            dosageInstructionFrequencyUnit = "",
            dosageInstructionFrequencyValue = 0.0,
            dosageInstructionMaxDosePerAdministration = 0.0,
            dosageInstructionPatientInstruction = "",
            dosageInstructionAsNeededDisplay = "",
            dosageInstructionPeriodBetweenDosagesUnit = "",
            dosageInstructionPeriodBetweenDosagesValue = 0.0,
            dosageDoseValue = "",
            dosageRateQuantityUnit = "",
            dosageDoseUnitDisplayOriginal = "",
            stopTypeDisplay = ""
        )
    }

    fun createTestLabEntries(): List<LabEntry> {
        return listOf(
            LabEntry(
                subject = TEST_SUBJECT,
                valueQuantityComparator = "",
                codeCodeOriginal = "LAB1",
                codeDisplayOriginal = "Lab Value 1",
                valueQuantityValue = 30.0,
                valueQuantityUnit = "U/l",
                referenceRangeText = "20 - 40",
                effectiveDateTime = LocalDate.of(2018, 5, 29),
            ), LabEntry(
                subject = TEST_SUBJECT,
                valueQuantityComparator = "",
                codeCodeOriginal = "LAB2",
                codeDisplayOriginal = "Lab Value 2",
                valueQuantityValue = 22.0,
                valueQuantityUnit = "mmol/L",
                referenceRangeText = "> 30",
                effectiveDateTime = LocalDate.of(2018, 5, 29),
            ), LabEntry(
                subject = TEST_SUBJECT,
                codeCodeOriginal = "LAB3",
                codeDisplayOriginal = "Lab Value 3",
                valueQuantityComparator = ">",
                valueQuantityValue = 50.0,
                valueQuantityUnit = "mL/min",
                referenceRangeText = "> 50",
                effectiveDateTime = LocalDate.of(2018, 5, 29),
            ), LabEntry(
                subject = TEST_SUBJECT,
                valueQuantityComparator = "",
                codeCodeOriginal = "LAB4",
                codeDisplayOriginal = "Lab Value 4",
                valueQuantityValue = 20.0,
                valueQuantityUnit = "mL/min",
                referenceRangeText = "",
                effectiveDateTime = LocalDate.of(2018, 5, 29),
            )
        )
    }

    private fun createTestVitalFunctionEntries(): List<VitalFunctionEntry> {
        return listOf(
            VitalFunctionEntry(
                subject = TEST_SUBJECT,
                effectiveDateTime = LocalDateTime.of(2021, 2, 27, 12, 30, 0, 0),
                codeDisplayOriginal = "NIBP",
                componentCodeDisplay = "Systolic blood pressure",
                quantityUnit = "mmHg",
                quantityValue = 120.0

            ),
            VitalFunctionEntry(
                subject = TEST_SUBJECT,
                effectiveDateTime = LocalDateTime.of(2021, 2, 26, 12, 30, 0, 0),
                codeDisplayOriginal = "NIBP",
                componentCodeDisplay = "Systolic blood pressure",
                quantityUnit = "mmHg",
                quantityValue = 1200.0
            ),
            VitalFunctionEntry(
                subject = TEST_SUBJECT,
                effectiveDateTime = LocalDateTime.of(2021, 2, 27, 12, 30, 0, 0),
                codeDisplayOriginal = "NIBP",
                componentCodeDisplay = "Systolic blood pressure",
                quantityUnit = "mmHg",
                quantityValue = 120.0
            ),
            VitalFunctionEntry(
                subject = TEST_SUBJECT,
                effectiveDateTime = LocalDateTime.of(2021, 2, 27, 12, 30, 0, 0),
                codeDisplayOriginal = "NIBP",
                componentCodeDisplay = "Diastolic blood pressure",
                quantityUnit = "mmHg",
                quantityValue = 120.0,
            )
        )
    }

    private fun createTestIntoleranceEntries(): List<IntoleranceEntry> {
        return listOf(
            IntoleranceEntry(
                subject = TEST_SUBJECT,
                assertedDate = LocalDate.of(2018, 1, 1),
                category = "medication",
                categoryAllergyCategoryDisplay = "",
                clinicalStatus = "",
                verificationStatus = "",
                codeText = "pills",
                criticality = "unknown",
                isSideEffect = "allergy"
            )
        )
    }

    private fun createTestBodyWeightEntries(): List<BodyWeightEntry> {
        return listOf(
            BodyWeightEntry(
                subject = TEST_SUBJECT,
                valueQuantityValue = 58.1,
                valueQuantityUnit = "kilogram",
                effectiveDateTime = LocalDateTime.of(2018, 4, 5, 12, 30, 0)
            ), BodyWeightEntry(
                subject = TEST_SUBJECT,
                valueQuantityValue = 61.1,
                valueQuantityUnit = "kilogram",
                effectiveDateTime = LocalDateTime.of(2018, 5, 5, 12, 30, 0)
            ), BodyWeightEntry(
                subject = TEST_SUBJECT,
                valueQuantityValue = 61.1,
                valueQuantityUnit = "kilogram",
                effectiveDateTime = LocalDateTime.of(2018, 5, 5, 12, 30, 0)
            ), BodyWeightEntry(
                subject = TEST_SUBJECT,
                valueQuantityValue = 611.0,
                valueQuantityUnit = "kilogram",
                effectiveDateTime = LocalDateTime.of(2018, 5, 4, 12, 30, 0)
            )
        )
    }
}